File: dev/update_packages.py
#!/usr/bin/env python
# Copyright 2015 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

import os
import subprocess

FLUTTER_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

def update(directory):
    packages = sorted(os.listdir(directory))
    for package in packages:
        package_dir = os.path.join(directory, package)
        if os.path.isdir(package_dir):
            print 'Updating', package, '...'
            subprocess.check_call(['pub', 'get'], cwd=package_dir)

update(os.path.join(FLUTTER_ROOT, 'packages'))
update(os.path.join(FLUTTER_ROOT, 'examples'))

File: dev/update_packages.dart
#!/usr/bin/env dart
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:io';

final String binaryName = Platform.isWindows ? 'pub.bat' : 'pub';
int runPub(Directory directory, List<String> pubArgs) {
  int updateCount = 0;
  for (FileSystemEntity dir in directory.listSync()) {
    if (dir is Directory) {
      updateCount++;
      Stopwatch timer = new Stopwatch()..start();
      stdout.write("Updating ${dir.path}...");
      ProcessResult result = Process.runSync(
        binaryName,
        pubArgs,
        workingDirectory: dir.path
      );
      timer.stop();
      stdout.write(" (${timer.elapsedMilliseconds} ms)");
      if (result.exitCode != 0) {
        print("... failed with exit code ${result.exitCode}.");
        print(result.stdout);
        print(result.stderr);
      } else {
        stdout.write("\n");
      }
    }
  }
  return updateCount;
}

void main(List<String> arguments) {
  Stopwatch timer = new Stopwatch()..start();
  bool upgrade = arguments.length > 0 && arguments[0] == '--upgrade';
  String FLUTTER_ROOT = new File(Platform.script.toFilePath()).parent.parent.path;
  List<String> pubArgs = [ upgrade ? 'upgrade' : 'get' ];
  int count = 0;
  count += runPub(new Directory("$FLUTTER_ROOT/packages"), pubArgs);
  count += runPub(new Directory("$FLUTTER_ROOT/examples"), pubArgs);
  String command = "$binaryName ${pubArgs.join(' ')}";
  print("Ran \"$command\" $count times in ${timer.elapsedMilliseconds} ms");
}
